### PR TITLE
增加evaluate_map

### PR DESCRIPTION
**Pull request by chaishare**
在原有仓库的基础上，增加了计算map指标的脚本

### CODE REVIEW

**Comment by tengerye** [comment]
Hi, after filtering the predictions not in the ground truth, you need to divide the result by the number of classes, i.e., averaging.

The following is the code to validate:

```
import torch
from torchmetrics.detection.mean_ap import MeanAveragePrecision

# Initialize the Mean Average Precision (mAP) metric
map_metric = MeanAveragePrecision()

# Ground truth annotations (bounding boxes, labels, and areas)
ground_truths = [
    {
        "boxes": torch.tensor([[50, 50, 100, 100], [30, 30, 60, 60]]), # Two ground truth boxes
        "labels": torch.tensor([1, 2]),  # Class 1: lamb, Class 2: dog
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100], [30, 30, 60, 60]]), # Two ground truth boxes
        "labels": torch.tensor([1, 2]),  # Class 1: lamb, Class 2: dog
    }
]

# Predictions
predictions = [
    {
        "boxes": torch.tensor([[55, 55, 95, 95]]),  # One predicted box
        "scores": torch.tensor([0.9]),  # Confidence score
        "labels": torch.tensor([3]),  # Class 3: cat
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]), # Two ground truth boxes
        "scores": torch.tensor([0.9]),  # Confidence score
        "labels": torch.tensor([1]),  # Class 1: lamb, Class 2: dog
    }
]

# Update the metric with ground truths and predictions
map_metric.update(predictions, ground_truths)

# Compute mAP
result = map_metric.compute()

# Print the result
print("mAP Score:", result)
```

**Comment by chaishare** [comment]
我补充了一些case来证明我们不需要对过滤的类别做额外的处理
test_case1和test_case2是证明预测数据中的那些“gt不存在类别的框”不会对结果map产生影响
test_case3是证明极端情况下map值也不会是100%
import torch
from torchmetrics.detection.mean_ap import MeanAveragePrecision

#  ==================================  testcase1  map值为0.3333333 ===================================================
map_metric_case1 = MeanAveragePrecision(iou_thresholds=[0.5])

ground_truths = [
    {
        "boxes": torch.tensor([[50, 50, 100, 100], [30, 30, 60, 60]]), # Two ground truth boxes
        "labels": torch.tensor([1, 2]),  # Class 1: lamb, Class 2: dog
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100], [30, 30, 60, 60], [40, 40, 70, 70]]), # Two ground truth boxes
        "labels": torch.tensor([1, 2, 3]),  # Class 1: lamb, Class 2: dog
    }
]

# Predictions
predictions = [
    {
        "boxes": torch.tensor([[50, 50, 100, 100], [40, 40, 70, 70]]),  # One predicted box
        "scores": torch.tensor([0.9, 0.9]),  # Confidence score
        "labels": torch.tensor([1, 3]),  # Class 3: cat
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([1]),
    }
]

# Update the metric with ground truths and predictions
map_metric_case1.update(predictions, ground_truths)

# Compute mAP
result = map_metric_case1.compute()

# Print the result
print("mAP Score:", result)




#  ==================================  testcase2  map值为0.3333333 ===================================================
# 说明预测数据中那些“gt中没有类别的预测框”不会对结果map产生影响
map_metric_case2 = MeanAveragePrecision(iou_thresholds=[0.5])

ground_truths = [
    {
        "boxes": torch.tensor([[50, 50, 100, 100], [30, 30, 60, 60]]), # Two ground truth boxes
        "labels": torch.tensor([1, 2]),  # Class 1: lamb, Class 2: dog
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100], [30, 30, 60, 60], [40, 40, 70, 70]]),
        "labels": torch.tensor([1, 2, 3]),
    }
]

# Predictions
predictions = [
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([1]),
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([1]),
    }
]

# Update the metric with ground truths and predictions
map_metric_case2.update(predictions, ground_truths)

# Compute mAP
result = map_metric_case2.compute()

# Print the result
print("mAP Score:", result)



#  ==================================  testcase3  map值为0.25 ===================================================
# 说明这种极端情况下，map值也不会是100%，而是是通过已知类别的召回率来控制的
map_metric_case3 = MeanAveragePrecision(iou_thresholds=[0.5])

ground_truths = [
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "labels": torch.tensor([1]),
    },
    {
            "boxes": torch.tensor([[50, 50, 100, 100]]),
            "labels": torch.tensor([1]),
    },
    {
            "boxes": torch.tensor([[50, 50, 100, 100]]),
            "labels": torch.tensor([1]),
    },
    {
            "boxes": torch.tensor([[50, 50, 100, 100]]),
            "labels": torch.tensor([1]),
    }
]

# Predictions
predictions = [
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([3]),
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([3]),
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([3]),
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([1]),
    }
]

# Update the metric with ground truths and predictions
map_metric_case3.update(predictions, ground_truths)

# Compute mAP
result = map_metric_case3.compute()

# Print the result
print("mAP Score:", result)

**Comment by chaishare** [comment]
我补充了一些case来证明我们不需要对过滤的类别做额外的处理
test_case1和test_case2是证明预测数据中的那些“gt不存在类别的框”不会对结果map产生影响
test_case3是证明极端情况下map值也不会是100%

```python
import torch
from torchmetrics.detection.mean_ap import MeanAveragePrecision

#  ==================================  testcase1  map值为0.3333333 ===================================================
map_metric_case1 = MeanAveragePrecision(iou_thresholds=[0.5])

ground_truths = [
    {
        "boxes": torch.tensor([[50, 50, 100, 100], [30, 30, 60, 60]]), # Two ground truth boxes
        "labels": torch.tensor([1, 2]),  # Class 1: lamb, Class 2: dog
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100], [30, 30, 60, 60], [40, 40, 70, 70]]), # Two ground truth boxes
        "labels": torch.tensor([1, 2, 3]),  # Class 1: lamb, Class 2: dog
    }
]

# Predictions
predictions = [
    {
        "boxes": torch.tensor([[50, 50, 100, 100], [40, 40, 70, 70]]),  # One predicted box
        "scores": torch.tensor([0.9, 0.9]),  # Confidence score
        "labels": torch.tensor([1, 3]),  # Class 3: cat
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([1]),
    }
]

# Update the metric with ground truths and predictions
map_metric_case1.update(predictions, ground_truths)

# Compute mAP
result = map_metric_case1.compute()

# Print the result
print("mAP Score:", result)




#  ==================================  testcase2  map值为0.3333333 ===================================================
# 说明预测数据中那些“gt中没有类别的预测框”不会对结果map产生影响
map_metric_case2 = MeanAveragePrecision(iou_thresholds=[0.5])

ground_truths = [
    {
        "boxes": torch.tensor([[50, 50, 100, 100], [30, 30, 60, 60]]), # Two ground truth boxes
        "labels": torch.tensor([1, 2]),  # Class 1: lamb, Class 2: dog
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100], [30, 30, 60, 60], [40, 40, 70, 70]]),
        "labels": torch.tensor([1, 2, 3]),
    }
]

# Predictions
predictions = [
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([1]),
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([1]),
    }
]

# Update the metric with ground truths and predictions
map_metric_case2.update(predictions, ground_truths)

# Compute mAP
result = map_metric_case2.compute()

# Print the result
print("mAP Score:", result)



#  ==================================  testcase3  map值为0.25 ===================================================
# 说明这种极端情况下，map值也不会是100%，而是是通过已知类别的召回率来控制的
map_metric_case3 = MeanAveragePrecision(iou_thresholds=[0.5])

ground_truths = [
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "labels": torch.tensor([1]),
    },
    {
            "boxes": torch.tensor([[50, 50, 100, 100]]),
            "labels": torch.tensor([1]),
    },
    {
            "boxes": torch.tensor([[50, 50, 100, 100]]),
            "labels": torch.tensor([1]),
    },
    {
            "boxes": torch.tensor([[50, 50, 100, 100]]),
            "labels": torch.tensor([1]),
    }
]

# Predictions
predictions = [
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([3]),
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([3]),
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([3]),
    },
    {
        "boxes": torch.tensor([[50, 50, 100, 100]]),
        "scores": torch.tensor([0.9]),
        "labels": torch.tensor([1]),
    }
]

# Update the metric with ground truths and predictions
map_metric_case3.update(predictions, ground_truths)

# Compute mAP
result = map_metric_case3.compute()

# Print the result
print("mAP Score:", result)
